extensions/wikia/Captcha/Captcha.class.php

```diff
@@ -0,0 +1,623 @@
+	const MAX_BAD_LOGIN_ATTEMPTS = 3;
+
+	protected $action;
+
+	protected $trigger;
+
+	/** @var \Captcha\Module\BaseCaptcha */
+	protected $captcha;
+
+	public function __construct() {
+		$this->captcha = Factory\Module::getInstance();
+	}
+
+	/**
+	 * Inject whazawhoo
+	 *
+	 * @fixme if multiple thingies insert a header, could break
+	 * https://wikia-inc.atlassian.net/browse/SOC-289
+	 *
+	 * @param \HTMLForm $form
+	 *
+	 * @return bool whether to keep running callbacks
+	 */
+	public function injectEmailUser( &$form ) {
+		$wg = \F::app()->wg;
+		if ( $wg->CaptchaTriggers['sendemail'] ) {
+			if ( $wg->User->isAllowed( 'skipcaptcha' ) ) {
+				$this->log( "user group allows skipping captcha on email sending\n" );
+				return true;
+			}
+			$form->addFooterText(
+				"<div class='captcha'>" .
+				$wg->Out->parse( $this->captcha->getMessage( 'sendemail' ) ) .
+				$this->captcha->getForm() .
+				"</div>\n" );
+		}
+		return true;
+	}
+
+	/**
+	 * Inject whazawhoo
+	 *
+	 * @fixme if multiple thingies insert a header, could break
+	 * https://wikia-inc.atlassian.net/browse/SOC-289
+	 *
+	 * @param \QuickTemplate $template
+	 *
+	 * @return bool whether to keep running callbacks
+	 */
+	public function injectUserCreate( &$template ) {
+		$wg = \F::app()->wg;
+		if ( $wg->CaptchaTriggers['createaccount'] ) {
+			if ( $wg->User->isAllowed( 'skipcaptcha' ) ) {
+				$this->log( "user group allows skipping captcha on account creation\n" );
+				return true;
+			}
+
+			$message = '';
+			wfRunHooks( 'GetConfirmEditMessage', [ $this, &$message ] );
+			if ( empty( $message ) ) {
+				$message = $this->captcha->getMessage( 'createaccount' );
+			}
+			$template->set( 'captcha',
+				"<div class='captcha'>" .
+				$this->captcha->getForm() .
+				'<p class="captchadesc" >' . $message . '</p>' .
+				"</div>\n" );
+		}
+		return true;
+	}
+
+	/**
+	 * Inject a captcha into the user login form after a failed
+	 * password attempt as a speedbump for mass attacks.
+	 *
+	 * @fixme if multiple thingies insert a header, could break
+	 * https://wikia-inc.atlassian.net/browse/SOC-289
+	 *
+	 * @param \QuickTemplate $template
+	 *
+	 * @return bool whether to keep running callbacks
+	 */
+	public function injectUserLogin( &$template ) {
+		if ( $this->isBadLoginTriggered() ) {
+			$template->set( 'header',
+				"<div class='captcha'>" .
+				\F::app()->wg->Out->parse( $this->captcha->getMessage( 'badlogin' ) ) .
+				$this->captcha->getForm() .
+				"</div>\n" );
+		}
+		return true;
+	}
+
+	/**
+	 * When a bad login attempt is made, increment an expiring counter
+	 * in the memcache cloud. Later checks for this may trigger a
+	 * captcha display to prevent too many hits from the same place.
+	 *
+	 * @param \User $user
+	 * @param string $password
+	 * @param int $retval authentication return value
+	 *
+	 * @return bool whether to keep running callbacks
+	 */
+	public function triggerUserLogin( $user, $password, $retval ) {
+		$wg = \F::app()->wg;
```

Comment:
You end up creating this local `$wg` variable in a lot of methods of this class. Thoughts on making it an instance variable that's initialized in the constructor?